### PR TITLE
Improve handling of speed changes

### PR DESCRIPTION
**Pull request by RomanBapst**
**Describe problem solved by this pull request**
It used to be that navigator handled VEHICLE_CMD_DO_CHANGE_SPEED and used the reposition triplet to propagate the changes to the controllers. This method has several drawback:
- not very efficient because the triplet needs to be republished for every speed change
- speed changes did not work for modes which do not rely on triplets being sent.
- republishing a triplet for the sake of updating cruise speed often lead to corner cases which had to be handled

**Describe your solution**
Allow the MC flightmodemanager and the FW position controllers to handle the commands directly. This allows speed changes to be implemented for any flight mode. Additionally, navigator does not need to republish the triplet every time a speed change occurs.
Navigator, will still save the newest speed setpoint and publish it as part of the next triplet update. This way speed changes persist mode changes.

In FW mode you can set the parameter `FW_POS_STK_CONF` to 0 which will disable airspeed setpoint via throttle stick (which is anyway weird if your throttle is spring loaded) and enable airspeed setpoints via command.

To test you have to build QGC master and use the **change speed** guided action.

**Test data / coverage**
Tested in SITL.

### CODE REVIEW

File: msg/vehicle_command.msg [hunk 1]
```diff
@@ -155,6 +155,11 @@ uint8 FAILURE_TYPE_SLOW = 5
 uint8 FAILURE_TYPE_DELAYED = 6
 uint8 FAILURE_TYPE_INTERMITTENT = 7
 
+uint8 SPEED_TYPE_AIRSPEED = 0
```
Comment (sfuhrer) [inline msg/vehicle_command.msg:158]:
```suggestion
uint8 DO_CHANGE_SPEED_TYPE_AIRSPEED = 0
```
To make it clear that it is linked to this command.

**Review by sfuhrer — COMMENTED** [review]
Having it directly in the FlightTask and the FW position controller removes some complexity and seems to make it cleaner :+1: 

Have some small comments. Plus, what do you think about also setting the ground speed in manual Position mode for MC to the value of the slider? It feels wrong that one can currently set the groundspeed also in Position mode for MC, but the value isn't actually used until one switches to an Auto mode.

File: msg/vehicle_command.msg (hunk 1 above)
Comment (RomanBapst) [reply msg/vehicle_command.msg:158]: @sfuhrer I'd rather introduce an enum in mavlink and name it accordingly, see e.g. https://mavlink.io/en/messages/ardupilotmega.html#SPEED_TYPE

**Comment by RomanBapst** [comment]
@dagar Mind taking a quick look and approving?

**Comment by sfuhrer** [comment]
>Plus, what do you think about also setting the ground speed in manual Position mode for MC to the value of the slider? It feels wrong that one can currently set the groundspeed also in Position mode for MC, but the value isn't actually used until one switches to an Auto mode

Thinking about this again, I think it's a bit dangerous to allow for setting the max speed in manual Position mode on MC. What if the user sets it to 1m/s, and then looses connection on the groundstation --> doesn't have the ability to increase it again, and would be forced to proceed flying at 1m/s. 
Rather we should probably disable the Set Speed slider on MC in any manual mode.

**Review by sfuhrer — DISMISSED** [review]

**Comment by MaEtUgR** [comment]
> In FW mode you can set the parameter FW_POS_STK_CONF to 0 which will disable airspeed setpoint via throttle stick (which is anyway weird if your throttle is spring loaded)

Maybe not directly related but maybe changing the airspeed with deflection would be useful like changing altitude in altitude mode (at least on MC)?

File: src/modules/flight_mode_manager/tasks/FlightTask/FlightTask.hpp [hunk 2]
```diff
@@ -175,6 +175,8 @@ class FlightTask : public ModuleParams
 		_acceleration_setpoint_feedback = acc_sp;
 	}
 
+	virtual bool setCruisingSpeed(const float cruising_speed_m_s) { return false; }
```
Comment (MaEtUgR) [inline src/modules/flight_mode_manager/tasks/FlightTask/FlightTask.hpp:178]:
```suggestion
	virtual bool setCruiseSpeed(const float cruise_speed_m_s) { return false; }
```
Just a suggestion.

File: src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.hpp [hunk 3]
```diff
@@ -92,6 +92,13 @@ class FlightTaskAuto : public FlightTask
 	bool updateInitialize() override;
 	bool update() override;
 
+	bool setCruisingSpeed(const float cruising_speed_m_s) override
+	{
+		_mc_cruise_speed = cruising_speed_m_s;
+		_commanded_speed_ts = hrt_absolute_time();
+		return true;
```
Comment (MaEtUgR) [inline src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.hpp:99]: I would make it void until there's actually a path that can fail and the return value gets handled.
```suggestion
```

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 4]
```diff
@@ -257,6 +257,18 @@ FixedwingPositionControl::vehicle_command_poll()
 
 				abort_landing(true);
 			}
+
+		} else if (vehicle_command.command == vehicle_command_s::VEHICLE_CMD_DO_CHANGE_SPEED
+			   && (uint8_t)vehicle_command.param1 == vehicle_command_s::SPEED_TYPE_AIRSPEED && vehicle_command.param2 > 0) {
```
Comment (MaEtUgR) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:262]: I would do the same as commented in the flight mode manager.

File: src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp [hunk 5]
```diff
@@ -344,8 +345,13 @@ bool FlightTaskAuto::_evaluateTriplets()
 
 	_type = (WaypointType)_sub_triplet_setpoint.get().current.type;
 
-	// Always update cruise speed since that can change without waypoint changes.
-	_mc_cruise_speed = _sub_triplet_setpoint.get().current.cruising_speed;
+	// Override a commanded cruise speed if the cruise speed from the triplet is valid and more recent
+	const float cruise_speed_from_triplet = _sub_triplet_setpoint.get().current.cruising_speed;
+
+	if (PX4_ISFINITE(cruise_speed_from_triplet) && cruise_speed_from_triplet >= 0
```
Comment (MaEtUgR) [inline src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp:351]:
```suggestion
	if (PX4_ISFINITE(cruise_speed_from_triplet) && cruise_speed_from_triplet > 0.f
```
Does a cruise speed == 0.f make sense and is it supported?

File: msg/vehicle_command.msg (hunk 1 above)
Comment (MaEtUgR) [inline msg/vehicle_command.msg:158]:
```suggestion
# param1 in DO_CHANGE_SPEED command
uint8 SPEED_TYPE_AIRSPEED = 0
```

**Review by MaEtUgR — COMMENTED** [review]
I found some comments. Mostly details and naming. If you're ok I'd just commit a suggestion and you can then decide and amend as you wish.

**Comment by MaEtUgR** [comment]
@RomanBapst @sfuhrer I rebased on master and added commits for my suggestions. Please feel free to adjust things as you like or tell me if I should change them.

File: src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.hpp [hunk 6]
```diff
@@ -201,6 +208,8 @@ class FlightTaskAuto : public FlightTask
 	_triplet_next_wp; /**< next triplet from navigator which may differ from the intenal one (_next_wp) depending on the vehicle state.*/
 	matrix::Vector3f _closest_pt; /**< closest point to the vehicle position on the line previous - target */
 
+	hrt_abstime _commanded_speed_ts{0};
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.hpp]: Changed in b866293750527dfae087aaf76a28a74b0190b4fd

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp [hunk 7]
```diff
@@ -266,6 +266,7 @@ class FixedwingPositionControl final : public ModuleBase<FixedwingPositionContro
 
 	float _manual_control_setpoint_altitude{0.0f};
 	float _manual_control_setpoint_airspeed{0.0f};
+	float _commanded_airspeed_setpoint{(float)NAN};		///< airspeed setpoint for manual modes commanded via MAV_CMD_DO_CHANGE_SPEED
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp]: Done in cdbb41acc703cc4777920551fc17de8a0bd0699b

File: src/modules/flight_mode_manager/FlightModeManager.cpp [hunk 8]
```diff
@@ -443,6 +443,13 @@ void FlightModeManager::handleCommand()
 			command_ack.target_component = command.source_component;
 			command_ack.timestamp = hrt_absolute_time();
 			_vehicle_command_ack_pub.publish(command_ack);
+
+		} else if (_current_task.task) {
+			// check for other commands not related to task switching
+			if (command.command == vehicle_command_s::VEHICLE_CMD_DO_CHANGE_SPEED && command.param2 >= 0
+			    && (uint8_t)command.param1 == vehicle_command_s::SPEED_TYPE_GROUNDSPEED) {
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/flight_mode_manager/FlightModeManager.cpp]: Done in cdbb41acc703cc4777920551fc17de8a0bd0699b

File: src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp (hunk 5 above)
Comment (sfuhrer) [reply src/modules/flight_mode_manager/tasks/Auto/FlightTaskAuto.cpp]: I think we should allow a very small speed, but probably not quite 0, so this change makes sense to me.

**Review by sfuhrer — APPROVED** [review]
Looks good to me :+1:

**Review by MaEtUgR — APPROVED** [review]

**Pull request merged**